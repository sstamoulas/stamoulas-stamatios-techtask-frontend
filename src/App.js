import React, { useEffect, useRef } from 'react'
import { connect } from 'react-redux'

import Main from './components/main/main.component'
import CustomLoader from './components/custom-loader/custom-loader.component'

import { fetchIngredientsStart } from './redux/ingredients/ingredients.actions'

import './App.styles.scss'

const App = ({ isFetching, fetchIngredientsStart }) => {
  const isMounted = useRef(false)

  useEffect(() => {
    if (!isMounted.current) {
      fetchIngredientsStart()
      isMounted.current = true
    }
  }, [isFetching, fetchIngredientsStart])

  return isMounted.current && !isFetching ?
    (
      <div className='App'>
        <Main />
      </div>
    )
  :
    <CustomLoader />
}

const mapStateToProps = (state) => ({
  isFetching: state.ingredients.isFetching,
})

const mapDispatchToProps = (dispatch) => ({
  fetchIngredientsStart: () => dispatch(fetchIngredientsStart()),
})

export default connect(mapStateToProps, mapDispatchToProps)(App)
