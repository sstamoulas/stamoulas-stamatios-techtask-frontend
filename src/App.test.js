import React from 'react'
import ReactDOM from 'react-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import { act } from 'react-dom/test-utils'
import { waitForElement, screen } from '@testing-library/react'

import App from './App'
import IngredientsActionTypes from './redux/ingredients/ingredients.types'

const mockStore = configureStore([])
const isFetching = true

describe('App Component', () => {
  let store
  let container
 
  beforeEach(() => {
    store = mockStore({
      ingredients: {
        isFetching,
      },
    })

    store.dispatch = jest.fn()

    container = document.createElement('div')
    document.body.appendChild(container)

    act(() => {
      ReactDOM.render(
        <Provider store={store}>
          <App />
        </Provider>, 
        container
      )
    })
  })

  it('should render with given state from Redux store', () => {
    expect(container).toMatchSnapshot()
  })

  it('should dispatch an action once the application has loaded', () => {
    expect(store.dispatch).toHaveBeenCalledTimes(1)
    expect(store.dispatch).toHaveBeenCalledWith({ 
      type: IngredientsActionTypes.FETCH_INGREDIENTS_START,
    })
  })
})
