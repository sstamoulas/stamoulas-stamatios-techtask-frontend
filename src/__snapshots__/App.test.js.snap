// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should render with given state from Redux store 1`] = `
<div>
  <svg
    style="height: 50vh; width: 50vw; display: block; margin: 10% auto;"
    viewBox="0 0 500 500"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M100 250 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.95;"
    />
    <path
      d="M106 208 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.9;"
    />
    <path
      d="M130 160 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.85;"
    />
    <path
      d="M160 130 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.8;"
    />
    <path
      d="M208 106 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.75;"
    />
    <path
      d="M250 100 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.7;"
    />
    <path
      d="M292 106 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.65;"
    />
    <path
      d="M340 130 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.6;"
    />
    <path
      d="M370 160 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.55;"
    />
    <path
      d="M394 208 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.5;"
    />
    <path
      d="M400 250 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.45;"
    />
    <path
      d="M394 292 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.4;"
    />
    <path
      d="M370 340 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.35;"
    />
    <path
      d="M340 370 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.3;"
    />
    <path
      d="M292 394 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.25;"
    />
    <path
      d="M250 400 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.2;"
    />
    <path
      d="M208 394 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.15;"
    />
    <path
      d="M160 370 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.1;"
    />
    <path
      d="M130 340 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0.05;"
    />
    <path
      d="M106 292 l3.668 7.568 8.332 1.151-6.064 5.828 1.48 8.279-7.416-3.967-7.417 3.967 1.481-8.279-6.064-5.828 8.332-1.151z"
      style="opacity: 0;"
    />
    <text
      dy=".5em"
      stroke="#000"
      stroke-width="2px"
      style="font-size: 40px;"
      text-anchor="middle"
      x="50%"
      y="50%"
    >
      Loading
    </text>
    <text
      dy=".5em"
      stroke="#000"
      stroke-width="2px"
      style="font-size: 40px;"
      text-anchor="middle"
      x="66%"
      y="50%"
    />
  </svg>
</div>
`;
